src/org/openstreetmap/josm/actions/mapmode/SplitMode.java

```diff
@@ -0,0 +1,352 @@
+                updateKeyModifiers(actionEvent);
+                if (platformMenuShortcutKeyMask) {
+                    JMenuItem source = (JMenuItem) actionEvent.getSource();
+                    JPopupMenu popup = (JPopupMenu) source.getParent();
+                    popup.remove(source);
+
+                    // Close popup menu anyway when there are no more options left
+                    if (popup.getSubElements().length > 0) {
+                        popup.setVisible(true);                                    
+                    }
+                }
+            });
+
+            pm.add(mi);
+        }
+        
+        MenuScroller.setScrollerFor(pm);
+        return pm;
+    }
+
+    /**
+     * Determine objects to highlight and update highlight
+     * @param e {@link MouseEvent} that triggered the update
+     * @return true if repaint is required
+     */
+    private boolean updateUserFeedback(MouseEvent e) {
+        List<OsmPrimitive> toHighlight = new ArrayList<>(2);
+
+        Optional<OsmPrimitive> pHovered = getPrimitiveAtPoint(e.getPoint());
+        DataSet ds = getLayerManager().getEditDataSet();
+
+        if (pHovered.filter(p -> p instanceof Node).isPresent()) {
+            Node nHovered = (Node) pHovered.get();
+            final List<Way> selectedWays = ds != null ? new ArrayList<>(ds.getSelectedWays()) : null;
+            List<Way> applicableWays = getApplicableWays(nHovered, selectedWays);
+            if (!applicableWays.isEmpty()) {
+                pHovered.ifPresent(toHighlight::add);
+            }
+            if (applicableWays.size() == 1) {
+                toHighlight.add(applicableWays.get(0));
+            }
+        }
+
+        return highlight.highlightOnly(toHighlight);
+    }
+
+    /**
+     * Removes all existing highlights.
+     * @return true if a repaint is required
+     */
+    private boolean removeHighlighting() {
+        boolean anyHighlighted = highlight.anyHighlighted();
+        highlight.clear();
+        return anyHighlighted;
+    }
+
+    /**
+     * Add a mouse listener to the component {@code c} which highlights {@code prims} 
+     * when the mouse pointer is hovering over the component
+     * @param c The component to add the hover mouse listener to
+     * @param prims The primitives to highlight when the component is hovered
+     */
+    private void addHoverHighlightListener(Component c, Collection<OsmPrimitive> prims) {
+        c.addMouseListener(new MouseAdapter() {
+            @Override
+            public void mouseEntered(MouseEvent e) {
+                if (highlight.highlightOnly(prims)) {
+                    MainApplication.getMap().mapView.repaint();
+                }
+            }
+
+            @Override
+            public void mouseExited(MouseEvent e) {
+                if (removeHighlighting()) {
+                    MainApplication.getMap().mapView.repaint();
+                }
+            }
+        });
+    }
```

Comment:
What for? I don't see why we would want to highlight individual way segments here

Reply:
Or do you mean for the repaint check in `removeHighlighting()`? 
The reasons not to use the data set highlight functions directly are:
1. It breaks the abstraction of the HighlightHelper which is supposed to more or less handle everything related to highlighting for us
2. It breaks the isolation between multiple sources of highlighting since they are all thrown into one big pool (e.g. in the case of `removeHighlighting()` it could lead to false positives and unnecessary repaints when there wasn't actually any highlighting removed but there is highlighting from somewhere else in the DataSet)

Reply:
Letting `HighlightHelper.clear()` return a boolean indicating whether anything was removed would be cleaner (and is essentially what is emulated here), but I didn't know if that would cause any backwards compatibility problems.